### DISCUSSION TITLE
Error: EPERM: operation not permitted - windows - custom schema.prisma location and custom  generated client output

### DISCUSSION THREAD

**Discussion by vatoer** [prisma/prisma#23223] [Q&A] [answered]
## what to accomplish

build nextjs using prisma with custom location because I will need multiple database connection

try

```sh
pnpm build
```

## error

```sh
> next build

   ▲ Next.js 14.1.0
   - Environments: .env

   Creating an optimized production build ...
glob error [Error: EPERM: operation not permitted, scandir 'C:\Users\admin\Application Data'] {
  errno: -4048,
  code: 'EPERM',
  syscall: 'scandir',
  path: 'C:\\Users\\admin\\Application Data'
}
Failed to compile.

Error: EPERM: operation not permitted, scandir 'C:\Users\admin\Application Data'
```

## possible culprit

I'm using custom location

my schema.prisma location `./prisma/my-db/schema.prisma`

## default location success

if I use default schema.prisma location `./prisma/schema.prisma` and default generated client it builds successfully

## why 'C:\Users\admin\Application Data' not 'C:\Users\admin\AppData'

this location should be on XP 'C:\Users\admin\Application Data' and on Windows 11 should be 'C:\Users\admin\AppData'

## how to reproduce 

- install nextjs
- install prisma
- setup schema on custom location ./prisma/my-db/schema.prisma
- setup custom generated client  output  to  "./generated/client"

- generate prisma client
  
  ```sh
  pnpm prisma generate --schema=./prisma/my-db/schema.prisma
  ```

- try to build
  
  ```sh
  pnpm build
  ```

- current result

```sh
next build

   ▲ Next.js 14.1.0
   - Environments: .env

   Creating an optimized production build ...
glob error [Error: EPERM: operation not permitted, scandir 'C:\Users\admin\Application Data'] {
  errno: -4048,
  code: 'EPERM',
  syscall: 'scandir',
  path: 'C:\\Users\\admin\\Application Data'
}
Failed to compile.

Error: EPERM: operation not permitted, scandir 'C:\Users\admin\Application Data'


> Build failed because of webpack errors
 ELIFECYCLE  Command failed with exit code 1.
```

### Expected behavior (optional)

- expected result
  
  no error, build success

### Information about Prisma Schema, Client Queries and Environment (optional)

windows 11

- username : admin 
- userprofile folder : C:\Users\admin
- %appdata% : C:\Users\admin\AppData\Roaming

```prisma
generator client {
  provider = "prisma-client-js"
  output   = "./generated/client"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Category {
  id       String    @id @default(cuid())
  name     String
}
```

**Comment by nurul3101** [comment]
Hey @vatoer 👋 

This issue does not appear to be directly related to Prisma but rather to how Next.js or your environment handles file access permissions during the build process.

The error `EPERM: operation not permitted, scandir 'C:\Users\admin\Application Data'` suggests that the build process is attempting to access a directory for which it doesn't have the necessary permissions. This could be due to various reasons, including how the paths are resolved in your environment or a misconfiguration.

Can you ensure that your user account (or the account running the build process) has full read/write access to the relevant directories? This includes the custom Prisma directories you've created and any temporary directories that might be used during the build process.

**Reply by vatoer** [reply]
yes, I don't have permissions to 'C:\Users\admin\Application Data'. 

even if the build use temporary path, afaik , as I mentioned earlier it should be ' C:\Users\admin\AppData' because I'm using windows 11 right?.

my project is located at `D:\myApp` ,  
my prisma 

`D:\myApp\prisma` 

```tree
prisma
└───my-db
    ├───generated
    │   └───client
    │       └───runtime
    └───migrations
```
        

I have full control and access over this path

Is there any next custom build configuration to solve this problem?

strangely, 
I can build successfully if the location of the `schema.prisma` and `generated client output` are leave as default.

**Reply by ludralph** [reply]
Hi @vatoer 

> Is there any next custom build configuration to solve this problem? 

If you're working in a monorepo setup, you can try using this plugin. Add it to your next.config.js:
```js
const { PrismaPlugin } = require('@prisma/nextjs-monorepo-workaround-plugin')

module.exports = {
  webpack: (config, { isServer }) => {
    if (isServer) {
      config.plugins = [...config.plugins, new PrismaPlugin()]
    }
    return config
  },
}
```
This solution is mentioned in this [GitHub comment](https://github.com/prisma/prisma/issues/10433#issuecomment-1414839304).

**Comment by omedinapr** [comment]
I'm having the same problem. If I add a custom output folder and then include it I get this error. It does NOT happen when left on default output.

Version: ^5.16.1

PS: Also when trying to build the NextJS project. Project in "D:/project" and its trying to access my "ApplicationData" folder. Build only fails when including the generated files.

**Reply by shmelevsergei** [reply]
Did you somehow solve this problem?

**Reply by omedinapr** [reply]
> Did you somehow solve this problem?

Did not find a working solution.

**Reply by shmelevsergei** [reply]
My application was developed on disk D of a shared physical disk. As soon as I moved it to the system disk C, the problem immediately disappeared. I don't know why it appeared, but literally a week ago there was nothing like this...

**Comment by jwyoo** [comment]
change the output fold to `node_modules/@prisma/clients/your_client_name`

**Comment by karuppusamy-d** [comment]
Adding generated prisma output directory to `experimental.outputFileTracingIgnores` in next config.js solves the issue. 

next config.js
```js
experimental: {
    outputFileTracingIgnores: ["./generated/client/**/*"],
}
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;

**Reply by trongthienpc** [reply]
I have same issue, any body help please!

**Comment by TheEgid** [comment]
Same problem. Only "D:\\" in prisma path. If "C:\\" is ok.

**Reply by chk7964** [reply]
with me also

**Comment by AliSaleemHasan** [comment]
@TheEgid  Just move generated prisma files to node-modules
```prisma
generator client {
  provider = "prisma-client-js"
  output   = "../../node_modules/generated"
}
```
and for getting the client or types you can import from generated as follow:
```node js
import { PrismaClient } from "generated";
```

**Reply by FrancoViniciusP** [reply]
Thanks! It worked here!

**Reply by TheEgid** [reply]
Thanks. But not worked in dev mode in next.js! Prohibition node_modules folder path in dev mode.

**Reply by shabanhr** [reply]
thanks it works for me

**Comment by Kiki-Q** [comment]
Does anyone find the right way to solve it?I have the same problem

**Comment by Kiki-Q** [comment]
@vatoer have you find out it? how do you solve this problem eventually??I can't solve it

**Reply by chk7964** [reply]
use c drive in windows or use a devcontainer

**Reply by vatoer** [reply, marked as answer]
> @vatoer have you find out it? how do you solve this problem eventually??I can't solve it

In windows, I put the output to `node_modules `

```tree
├─node_modules
├─prisma
│   ├─db-1
│   │   └─schema.prisma
│   └─db-2
│        └─schema.prisma
│
├─public
├─src
    └─lib
       ├─db1.ts
       └─db1.ts
```

`prisma\db-1\schema.prisma`

```prisma
generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native", "debian-openssl-3.0.x"]
  output        = "../../node_modules/@prisma-db-1/client"
}
```



and then in `src\lib\db1.ts` I can Import 

```
import { Prisma, PrismaClient } from "@prisma-db-1/client";

// Extend the global object to include prismaDb1
declare global {
  // eslint-disable-next-line no-var
  var prismaDb1: PrismaClient | undefined;
}

const globalForPrisma = global as typeof global & {
  prismaDb1?: PrismaClient;
};

export const db1 = globalForPrisma.prismaDb1 || new PrismaClient();

if (process.env.NODE_ENV !== "production")
  globalForPrisma.prismaDb1 = db1;

export { Prisma };
```

so you can use it as 

```ts
import { db1 } from "@/lib/db1";

export const getRecords = async () = {
 const records = await db1.yourtable.findMany({})
}

```

**Reply by cupid20103** [reply]
This is no working on the production version. Is there any soft soltion?

**Reply by waltercorneli** [reply]
Something like this work for me but it throw out some eslint errors inside files in `generated` folders.
